src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala

```diff
@@ -74,43 +75,83 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
       // from within the batch, but we can only do this if the offset falls within our range of the minimum and
       // maximum offset we have across all the CastleMessageBatches within our data map
       if (offset >= minOffset && offset <= maxOffset)
-        // We iterate over all of the castle message batches we have in data and attempt to
-        // create a batch from the given offset if it happens to fall inside that batch
-        // createBatchFromOffset returns Some(CastleMessageBatch) if it is successful and None otherwise.
+      // We iterate over all of the castle message batches we have in data and attempt to
+      // create a batch from the given offset if it happens to fall inside that batch
+      // createBatchFromOffset returns Some(CastleMessageBatch) if it is successful and None otherwise.
         findOption(data.values)(castleMessageBatch => castleMessageBatch.createBatchFromOffset(offset))
       else
         None
     }
   }
 
   /**
-   * findOption iterates through an Iterable of A and applies function f to each value of type A
-   * which returns back an Option[B]
-   * If Option[B] is not empty findOption will return that as the result,
-   * Otherwise findOption continues going through the remaining values.
-   * If findOption exhausts all values without encounering a Some value, it returns None
-   */
+    * Gets all messages from the cache that fit in the specified buffer size starting with the provided offset.
+    * If there is a cache hit this method returns atleast one message irrespective of the bufferSize otherwise
+    * returns a None in case of a miss.
+    * @param offset
+    * @param bufferSize
+    * @return
+    */
+  def getAll(offset: Long, bufferSize: Int): Option[CastleMessageBatch] = {
+    get(offset).map{ batch =>
+      val messageBatches = getMoreRecursive(Vector(batch), bufferSize)
+      if(messageBatches.size > 1)
+        // Need to concatenate batches into a single batch
+        CastleMessageBatch(messageBatches)
+      else
+        messageBatches.head
+    }
+  }
+
+  /**
+    * Recursive function to keep fetching more batches from the cache with the nextOffset of the previous batch
+    * till a cache miss happens or we exhaust the buffer capacity.
+    * @param currentList
+    * @param bufferSize
+    * @return
+    */
   @tailrec
-  private def findOption[A,B](values: Iterable[A])(f: A => Option[B]): Option[B] =
-    values.headOption match {
-      case Some(h) =>
-        f(h) match {
-          case Some(v) => Some(v)
-          case None => findOption(values.tail)(f)
-        }
-      case None => None
+  private def getMoreRecursive(currentList: Vector[CastleMessageBatch], bufferSize: Int): Vector[CastleMessageBatch] = {
```

Comment:
Here we querying the cache with the nextOffset of the last batch. This will ensure that the batches are contiguous right ? If there is no batch  in the cache that starts with nextOffset of the last batch we return the list upto the last batch. Lemme know if am missing a corner case here.